src/test/java/org/opensearch/knn/jni/JNIServiceTests.java

```diff
@@ -700,19 +709,98 @@ public void testQueryIndex_faiss_valid() throws IOException {
+    public void testQueryIndex_faiss_parentIds() throws IOException {
+
+        int k = 100;
+
+        List<String> methods = ImmutableList.of(faissMethod);
+        List<SpaceType> spaces = ImmutableList.of(SpaceType.L2, SpaceType.INNER_PRODUCT);
+        int[] parentIds = toParentIdArray(testDataNested.indexData.docs);
+        Map<Integer, Integer> idToParentIdMap = toIdToParentIdMap(testDataNested.indexData.docs);
+        for (String method : methods) {
+            for (SpaceType spaceType : spaces) {
+                Path tmpFile = createTempFile();
+                JNIService.createIndex(
+                    testDataNested.indexData.docs,
+                    testDataNested.indexData.vectors,
+                    tmpFile.toAbsolutePath().toString(),
+                    ImmutableMap.of(INDEX_DESCRIPTION_PARAMETER, method, KNNConstants.SPACE_TYPE, spaceType.getValue()),
+                    FAISS_NAME
+                );
+                assertTrue(tmpFile.toFile().length() > 0);
+
+                long pointer = JNIService.loadIndex(
+                    tmpFile.toAbsolutePath().toString(),
+                    ImmutableMap.of(KNNConstants.SPACE_TYPE, spaceType.getValue()),
+                    FAISS_NAME
+                );
+                assertNotEquals(0, pointer);
+
+                for (float[] query : testDataNested.queries) {
+                    KNNQueryResult[] results = JNIService.queryIndex(pointer, query, k, FAISS_NAME, null, parentIds);
+                    // Verify there is no more than one result from same parent
+                    Set<Integer> parentIdSet = toParentIdSet(results, idToParentIdMap);
+                    assertEquals(results.length, parentIdSet.size());
+                }
+            }
+        }
+    }
+
+    private Set<Integer> toParentIdSet(KNNQueryResult[] results, Map<Integer, Integer> idToParentIdMap) {
+        Set<Integer> parentIdSet = new HashSet<>();
+        for (KNNQueryResult result : results) {
+            parentIdSet.add(idToParentIdMap.get(result.getId()));
+        }
+        return parentIdSet;
+    }
+
+    private int[] toParentIdArray(int[] ids) {
+        int length = ids.length;
+        int[] sortedIds = Arrays.copyOf(ids, length);
+        Arrays.sort(sortedIds);
+
+        List<Integer> parentIds = new ArrayList<>();
+        parentIds.add(sortedIds[length - 1] + 1);
+        for (int i = length - 2; i > -1; i--) {
+            if (sortedIds[i] != sortedIds[i + 1] - 1) {
+                parentIds.add(sortedIds[i] + 1);
+            }
+        }
+
+        Collections.shuffle(parentIds);
+        return parentIds.stream().mapToInt(Integer::intValue).toArray();
+    }
+
+    private Map<Integer, Integer> toIdToParentIdMap(int[] ids) {
+        int length = ids.length;
+        int[] sortedIds = Arrays.copyOf(ids, length);
+        Arrays.sort(sortedIds);
+
+        Map<Integer, Integer> idToParentIdMap = new HashMap<>();
+        int parentId = sortedIds[length - 1] + 1;
```

Comment:
can you please extract last element of array into it's own variable with meaningful name, for instance `lastId` or 'largestId`? I see it's used at least 3 times, doing that should increase code readability.

Reply:
It is not the last id. It is parent id of current id.